### ISSUE TITLE
[Ask for advice] General machine learning or tensorflow first

### ISSUE THREAD

**Issue by aunnnn** [off99555/machine-learning-curriculum#4]
Would you recommend diving into tensorflow applications first without learning about general machine learning algorithms? 

E.g. Should I took Andrew's ML course or Kadenze's creative applications with Tensorflow first?

Thanks,
Aunn

**Comment by offchan42** [comment]
I think that Kadenze is too difficult for you if you don't understand ML ideas before.
To test that hypothesis, try watching a couple of lectures. If you aren't confused then congrats.
But if you are confused, please consider applying machine learning in general first. Either Intro to Machine Learning at Udacity or at Coursera works. 

===

They will teach you the concept of overfitting (variance), underfitting (bias), generalization, regularization, normalization, validation, testing metrics (accuracy, precision, recall, cost/loss, r-squared), feature extraction, curse of dimensionality, dimensionality reduction (PCA, Manifold Learning, t-SNE), learning tasks (regression, classification), optimization techniques, gradient descent, data visualization, maths, etc. All those fun kinds of stuff.

**Comment by offchan42** [comment]
Coursera is for thorough math introduction. Udacity is for intuition and practical usage.

**Comment by offchan42** [comment]
Personally, I completed all of them. Intro at Udacity first then Andrew Ng at Coursera. Then you can pursue Deep Learning at Udacity and Kadenze. All of them are recommended in the list.

Update: Now deep learning is also taught at coursera as a specialization, so you can also take this course because I think it covers everything from the fundamentals.

**Comment by aunnnn** [comment]
That's a very useful insight. Thank you very much!

**Comment by offchan42** [comment]
**Update in 2020:**
I still suggest Andrew Mg course as the best introduction to machine learning. Then for deep learning, Andrew Ng also has a course at coursera. I believe in his teaching quality so I suggest you to take the course. (Though I haven't completed the deep learning course by Andrew as I already have the deep learning knowledge from other courses).

**Comment by offchan42** [comment]
**Update in 2021:**
Python has taken over the data science and machine learning world completely but Andrew Ng's machine learning course was taught using Octave language. I find this inefficient. But there's a repo which releases all course assignments in python instead of Octave so I think you should use it: https://github.com/dibgerge/ml-coursera-python-assignments

Another course you could try is this: https://www.coursera.org/learn/machine-learning-with-python

**Comment by offchan42** [comment]
**Update in 2022:**
Andrew Ng has released a new course specialization with focus on python instead of Octave, so you could take it instead here: https://www.coursera.org/specializations/machine-learning-introduction

**Comment by offchan42** [comment]
**Update in 2023:**
The last year latent diffusion models and GPT-3 have been occupying the space. Transformer architecture is killing it.
This year I suggest mostly fast.ai courses for practical purposes, Andrew Ng's specialization for theoretical understanding, and Andrej Karpathy videos on his YouTube channel for hands-on implementing things from scratch for bottom-up understanding.
